portable/ARMv8M/non_secure/port.c

```diff
@@ -568,34 +579,61 @@ PRIVILEGED_DATA static volatile uint32_t ulCriticalNesting = 0xaaaaaaaaUL;
+             * use the core clock. This configuration forces the SysTick to load
              * from portNVIC_SYSTICK_LOAD_REG immediately instead of at the next
-             * cycle of the other clock.  Then portNVIC_SYSTICK_LOAD_REG is ready
-             * to receive the standard value immediately. */
+             * cycle of the external clock. Then portNVIC_SYSTICK_LOAD_REG is ready
+             * to receive the standard value immediately.
+             */
+
             portNVIC_SYSTICK_CURRENT_VALUE_REG = 0UL;
-            portNVIC_SYSTICK_CTRL_REG = portNVIC_SYSTICK_CLK_BIT | portNVIC_SYSTICK_INT_BIT | portNVIC_SYSTICK_ENABLE_BIT;
+
             #if ( portNVIC_SYSTICK_CLK_BIT_CONFIG == portNVIC_SYSTICK_CLK_BIT )
             {
                 portNVIC_SYSTICK_LOAD_REG = ulTimerCountsForOneTick - 1UL;
+
+                /* Enable SysTick */
+                portNVIC_SYSTICK_CTRL_REG |= portNVIC_SYSTICK_ENABLE_BIT;
             }
             #else
             {
-                /* The temporary usage of the core clock has served its purpose,
-                 * as described above.  Resume usage of the other clock. */
-                portNVIC_SYSTICK_CTRL_REG = portNVIC_SYSTICK_CLK_BIT | portNVIC_SYSTICK_INT_BIT;
+                /* Temporarily switch to core clock */
+                ulCtrlRegValue |= portNVIC_SYSTICK_CLK_BIT;
+                portNVIC_SYSTICK_CTRL_REG = ulCtrlRegValue;
+
+                /* Enable SysTick */
+                ulCtrlRegValue |= portNVIC_SYSTICK_ENABLE_BIT;
+                portNVIC_SYSTICK_CTRL_REG = ulCtrlRegValue;
+
+                /* Read portNVIC_SYSTICK_CTRL_REG into ulCtrlRegValue */
+                ulCtrlRegValue = portNVIC_SYSTICK_CTRL_REG;
 
-                if( ( portNVIC_SYSTICK_CTRL_REG & portNVIC_SYSTICK_COUNT_FLAG_BIT ) != 0 )
+                /* Disable SysTick */
+                ulCtrlRegValue &= ~( portNVIC_SYSTICK_ENABLE_BIT );
+                portNVIC_SYSTICK_CTRL_REG = ulCtrlRegValue;
+
+                /* Update ulCtrlRegValue */
+                ulCtrlRegValue |= portNVIC_SYSTICK_CTRL_REG;
+
+                /* Switch back to external clock. */
+                ulCtrlRegValue &= ~( portNVIC_SYSTICK_CLK_BIT );
+                portNVIC_SYSTICK_CTRL_REG = ulCtrlRegValue;
+
+                if( ( ulCtrlRegValue & portNVIC_SYSTICK_COUNT_FLAG_BIT ) != 0 )
                 {
                     /* The partial tick period already ended.  Be sure the SysTick
                      * counts it only once. */
                     portNVIC_SYSTICK_CURRENT_VALUE_REG = 0;
                 }
 
                 portNVIC_SYSTICK_LOAD_REG = ulTimerCountsForOneTick - 1UL;
-                portNVIC_SYSTICK_CTRL_REG = portNVIC_SYSTICK_CLK_BIT_CONFIG | portNVIC_SYSTICK_INT_BIT | portNVIC_SYSTICK_ENABLE_BIT;
+
+                /* Enable SysTick */
+                ulCtrlRegValue |= ( portNVIC_SYSTICK_ENABLE_BIT | portNVIC_SYSTICK_INT_BIT );
```

Comment:
Code sequence needs considering carefully as it's potentially using an outdated ulCtrlRegValue value.